src/Search.jl

```diff
@@ -0,0 +1,110 @@
+export pkgsearch
+
+dir(token) = joinpath(homedir(),".julia","v$(VERSION.major).$(VERSION.minor)",token)
+
+# recursive function for applying search criteria
+function keycheck!(data::Dict{<:Any,<:Any},str::Array{String,1},mode::Array{Symbol,1},uid::String)
+    found = false
+    for key in data |> keys |> collect |> sort
+        if :deps in mode
+            for s in str
+                (key == s) && (found = true)
+            end
+        end
+        if (((:search in mode) && (typeof(data[key]) == String)) ||
+            ((:name in mode) && (key == "name")) ||
+            ((:desc in mode) && (key == "description")))
+            for s in lowercase.(str)
+                occursin(s,lowercase(data[key])) && (found = true)
+            end
+        end
+        if (:keywords in mode) && (key == "keywords")
+            for s in str
+                (s in data[key]) && (found = true)
+            end
+        end
+        (key == "uuid") && if (:uuid in mode)
+            for s in str
+                (data[key] == s) && (found = true) && (uid = s)
+            end
+        else
+            found && (uid = data[key])
+        end
+        if typeof(data[key]) <: Dict{<:Any,<:Any}
+            (fd, uid) = keycheck!(data[key],str,mode,uid) 
+            fd && (found = true)
+        end
+    end
+    return (found, uid)
+end
+
+# directory search for package toml
+function pkgsearch(mode::Array{Symbol,1},str::Array{String,1})
+    m = mode
+    (:search in m) && push!(m,:deps,:name,:keywords)
+    (:uuid in m) && (:name ∉ m) && push!(m,:name)
+    pkglist = Pair{String,String}[]
+    path = joinpath(homedir(),".julia","registries")
+    for depot in readdir(path)
+        data = nothing
+        for (root, dirs, files) in walkdir(joinpath(path,depot))
+            for dir in dirs
+                found = false
+                uid = ""
+                for file in readdir(joinpath(root,dir))
+                    if endswith(file, ".toml")
+                        (file == "Deps.toml") && (:deps ∉ m) && continue
+                        (file == "Package.toml") && (:name ∉ m) && continue
+                        (file != "Project.toml") && (:deps ∉ m) && (:name ∉ m) && continue
+                        data = TOML.parsefile(joinpath(root, dir, file))
```

Comment:
Searching through and parsing all the TOML files in a registry does not seem scalable. This would take a long time already and will take even longer as more packages are registered.

Reply:
It's actually very fast, see https://github.com/JuliaLang/Pkg3.jl/pull/116#issue-167638105